### PR TITLE
Introduce dot for install and build

### PR DESCRIPTION
**Pull request by bronzdoc**
# Description:
closes https://github.com/rubygems/rubygems/issues/2580

This PR introduce the use of the `.` comand line argument to use when building, installing or uninstalling gems in the current directory.

The `.` argument will build all `.gemspec` files when building i.e `gem build .` and it will install/uninstall all the `.gem` files when installing or uninstalling i.e `gem install .` or `gem uninstall .`

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by MSP-Greg** [comment]
If one has gems a, b, c, & d in the 'dot' directory, and d is a dependency of a, will all four gems be installed?  Or, does `gem install .` look to rubygems.org for d?

If d is not installed from '.', should dependencies first be looked for in the 'dot' directory?

**Comment by bronzdoc** [comment]
@MSP-Greg if one has gems a, b, c in the 'dot' directory, and you want to install them as of now you need to to `gem install` 3 times, this will make you just have to do `gem install` one time, i'm not changing how `gem install` works.

**Comment by MSP-Greg** [comment]
@bronzdoc

Yes, I understand the change.  My question involves how dependencies of a, b, c are handled, in terms of whether they are also installed from the 'dot' folder, assuming they exist there...

**Comment by MSP-Greg** [comment]
@bronzdoc

Never mind.  An edge case like this leads to a deep rabbit hole without a non-trival solution.  If a dependency in dot exists, one should use the `-​-ignore-dependencies` option...

**Comment by bronzdoc** [comment]
@MSP-Greg see https://github.com/rubygems/rubygems/issues/448

**Comment by MSP-Greg** [comment]
@bronzdoc 

Thanks for the issue link.  I like the idea in @deivid-rodriguez's post: https://github.com/rubygems/rubygems/issues/448#issuecomment-364673650.  I can't look at it until later, does this PR accomplish that also?

**Comment by bronzdoc** [comment]
@MSP-Greg 
>  does this PR accomplish that also?

nope this is not changing `gem install` behavior :)

**Comment by MSP-Greg** [comment]
@bronzdoc 

> nope this is not changing gem install behavior :)

Ok.  

1. I haven't ever needed to install more that one local gem, so I've never checked to see if one could load several from a hard code path parameter.

2. If RG `gem install` didn't previously respond to `gem install .`, then previous behavior has not been defined, so we're free to do whatever.

3. If that behavior differs from similar options (like the first point), then that is something that should be discussed.

Kind of like `gem update --explain`  :)

**Comment by deivid-rodriguez** [comment]
Having several gemspecs in the same directory seems... a really weird edge case. I'm not sure whether we should support that, since nobody has requested it? @voxik said he _might_ need to do that if https://github.com/rubygems/rubygems/issues/2587 was not resolved, but #2587 has been resolved.

I think I would prefer for now to build the single gemspec in the directory and fail with a message telling the user to be explicit if there are actually multiple gemspecs?

**Comment by bronzdoc** [comment]
@deivid-rodriguez 

@LemonAndroid requested it
> @djberg96 @voxik I think it should just build or install for all gemspecs by default.

@LemonAndroid can you be more explicit why would you need to build all gemspecs, trying to understand the edge case and see if it is worth keeping this PR the way it is or modify it with what @deivid-rodriguez suggested, thanks!

File: test/rubygems/test_gem_commands_build_command.rb [hunk 1]
```diff
@@ -230,6 +230,90 @@ def test_execute_outside_dir
...
+  def test_execute_current_dir
+    gem_0 = util_spec 'some_gem' do |s|
+      s.license = 'AGPL-3.0'
+      s.files = ['README.md']
+    end
+
+    gem_1 = util_spec 'another_gem' do |s|
+      s.license = 'AGPL-3.0'
+      s.files = ['README.md']
+    end
```
Comment (deivid-rodriguez) [inline test/rubygems/test_gem_commands_build_command.rb]: My fear is that supporting this might cause confusion and it's hard to think of cases where it's useful. For example, in this test we're defining two different gems that have the same README... right? :thinking:

**Comment by deivid-rodriguez** [comment]
Thanks @bronzdoc, I missed that sentence when reading the issue. :+1:

**Comment by deivid-rodriguez** [comment]
The thing is... Having multiple gemspecs in the same folder is... weird to say the least, because either `spec.files` in all gemspecs will be pointing to the same files (but why are the gems diffferent then?), or `spec.files` will list disjoint sets of files (but in that case, wouldn't moving each gem to a subfolder be way clearer?).

I think we should at least have a realworld example of this before we consider supporting it.

**Comment by LemonAndroid** [comment]
> The thing is... Having multiple gemspecs in the same folder is... weird to say the least, because either `spec.files` in all gemspecs will be pointing to the same files (but why are the gems diffferent then?), or `spec.files` will list disjoint sets of files (but in that case, wouldn't moving each gem to a subfolder be way clearer?).
> 
> I think we should at least have a realworld example of this before we consider supporting it.

There could be multiple folders + the .gemspec file. Each folder containing the files related to one .gemspec file.

**Comment by deivid-rodriguez** [comment]
Aha, that makes sense and I have seen that case, but in my case this wouldn't work because all gems depend on each other, so the "isolated installs" would fail. I would need something like https://github.com/rubygems/rubygems/issues/448 instead.

In any case, I don't think that's what this PR is implementing? I think it's only look for gems in the root folder...

**Comment by LemonAndroid** [comment]
> #448

It should basically get all gemspecs from a folder + sub folders and resolve dependencies automatically?

**Comment by deivid-rodriguez** [comment]
Yes, that's what I requested there because I've seen situations where I think it could be useful. For example, https://github.com/rails/rails or https://github.com/decidim/decidim.

**Comment by bronzdoc** [comment]
Ok, for now lets keep this simple, since https://github.com/rubygems/rubygems/issues/448 is not yet implemented and there's no other real world usage example,  i'll change this PR to what @deivid-rodriguez suggested.

Any objection on the way `gem install .` will work?

**Comment by LemonAndroid** [comment]
Nope

**Comment by voxik** [comment]
From top of my head, here are several examples of libraries with more than one .gemspec in directory:

https://github.com/ruby-concurrency/concurrent-ruby
https://github.com/flori/json/

I am pretty sure there is more. Typically it is useful for gems which are plain Ruby, but they can benefit from binary extension as well. I can imagine, that there might be cases of multiple .gemspec in one repository for better multiplatform support, but I don't have example at my hand. And the concurrent-ruby example is definitely very interesting.

And my use case is a bit specific. If #2587 was not resolve, we would need to use workflow like this [[1]]

~~~
$ gem unpack %{SOURCE0}
$ gem spec %{SOURCE0} -l --ruby > %{gem_name}.gemspec

$ cat some.patch | patch -p1
$ sed -i "s/modify/something/" somewhere/in/the/package

$ gem build %{gem_name}.gemspec
~~~

But there is quite a lot of gems shipping their upstream .gemspec, which might collide with the generated `%{gem_name}.gemspec`.

Luckily, because I can use `gem build ../foo.gemspec` again, I don't have to place two .gemspec files into one directory for my purposes and we follow a bit different set of guidelines [[2]].

[1]: https://fedoraproject.org/w/index.php?title=Packaging:Ruby&oldid=400800#Building_gems
[2]: https://docs.fedoraproject.org/en-US/packaging-guidelines/Ruby/#_building_gems

**Comment by deivid-rodriguez** [comment]
Thanks @voxik for the explanation! So this is an improvement you would welcome, then? If this can be useful for gems out there, I guess we can ship it?

We could improve the patch in the future by supporting arbitrary folder instead of only ".", but this is a good start :+1:.

**Comment by voxik** [comment]
I guess that `gem install some/path/*.gem` already works just fine if I am not mistaken. `gem install .` is ambiguous and magical to me, not mentioning that this PR add it just as special case for current directory.

It seems that `gem build *.gemspec` is not supported, so this could be probably nice extension, but I have never had any need for this myself.

If you are talking about cases such as Rails repository, I still think there is nice place to use shell `find`, but if you want to really support something like recursive lookup, then there should be probably some `-R` flag, which is used by `grep`, `chown` and similar.
 
Frankly, reading the #2580, I think it should be rejected as it is and this should be started over with refined scope.

**Comment by deivid-rodriguez** [comment]
@voxik I agree with what you say, I think adding support for `gem build *.gemspec` satisfies the original request quite well and it's very intuitive. I support we do that instead :+1:.

**Comment by bronzdoc** [comment]
I'm happy to work on this if we can all agree in the new scope :)

**Comment by deivid-rodriguez** [comment]
Hahaha, right, I'm sorry @bronzdoc, but @voxik raised good points and made me change my mind.

I think the most intuitive addition is indeed supporting multiple arguments to `gem install` and `gem build`, and looping over them the same way this PR is doing over the files in the current directory.
